File: python/go_counting.py
from collections import defaultdict

WHITE = 0
BLACK = 1
NONE = 2


class UnionFind:
    def __init__(self):
        self.root = {}
        self.rank = defaultdict(int)

    def find(self, k):
        if k not in self.root:
            self.root[k] = k
            return k

        root = self.root[k]
        if root == k:
            return k
        root = self.find(root)
        self.root[k] = root
        return root

    def union(self, a, b):
        ra = self.find(a)
        rb = self.find(b)
        if ra == rb:
            return

        if self.rank[ra] < self.rank[rb]:
            self.root[ra] = rb
        elif self.rank[ra] > self.rank[rb]:
            self.root[rb] = ra
        else:
            self.root[rb] = ra
            self.rank[ra] += 1

    def to_groups(self):
        groups = defaultdict(set)
        for k, root in self.root.items():
            groups[root].add(k)
        return groups


class Board:
    def __init__(self, board):
        self.board = board
        self.uf = UnionFind()
        self.row = len(board)
        self.col = len(board[0])
        for i in range(self.row):
            for j in range(self.col):
                if board[i][j] != " ":
                    continue
                if i > 0 and board[i - 1][j] == " ":
                    self.uf.union((j, i - 1), (j, i))
                if j > 0 and board[i][j - 1] == " ":
                    self.uf.union((j - 1, i), (j, i))
                self.uf.union((j, i), (j, i))
        self.groups = self.uf.to_groups()
        self.border = {}
        self.check_border()

    def check_border(self):
        dir = [(-1, 0), (1, 0), (0, 1), (0, -1)]
        for k, g in self.groups.items():
            isblack = False
            iswhite = False
            for j, i in g:
                for dj, di in dir:
                    ni = i + di
                    nj = j + dj
                    if not self.is_valid_coor(ni, nj):
                        continue
                    if self.board[ni][nj] == "B":
                        isblack = True
                    elif self.board[ni][nj] == "W":
                        iswhite = True

            if isblack and not iswhite:
                self.border[k] = BLACK
            elif iswhite and not isblack:
                self.border[k] = WHITE
            else:
                self.border[k] = NONE

    def is_valid_coor(self, i, j):
        return i >= 0 and i < self.row and j >= 0 and j < self.col

    def territory(self, x, y):
        if not self.is_valid_coor(y, x):
            raise ValueError("Invalid coordinate")

        if self.board[y][x] != " ":
            return NONE, set()

        root = self.uf.find((x, y))

        return self.border[root], self.groups[root]

    def territories(self):
        territories = defaultdict(set)
        for k, v in self.border.items():
            territories[v] |= self.groups[k]
        return territories

File: elixir/go_counting.ex
defmodule GoCounting do
  @type position :: {integer, integer}
  @type owner :: %{owner: atom, territory: [position]}
  @type territories :: %{white: [position], black: [position], none: [position]}

  @doc """
  Return the owner and territory around a position
  """
  @spec territory(board :: String.t(), position :: position) ::
          {:ok, owner} | {:error, String.t()}
  def territory(board, {x, y} = pos) do
    g = to_grid(board)
    {c, r} = g |> Map.keys() |> Enum.max()

    cond do
      x < 0 or y < 0 or y > c or x > r ->
        {:error, "Invalid coordinate"}

      g[{x, y}] != "_" ->
        {:ok, %{owner: :none, territory: []}}

      true ->
        g
        |> to_groups()
        |> Enum.find(&(pos in &1))
        |> then(fn t ->
          {:ok, %{owner: check(t, g), territory: t}}
        end)
    end
  end

  defp check(lst, g) do
    lst
    |> Enum.reduce({0, 0}, fn {x, y}, acc ->
      [{-1, 0}, {0, 1}, {1, 0}, {0, -1}]
      |> Enum.reduce(acc, fn {dx, dy}, {b, w} ->
        case g[{x + dx, y + dy}] do
          "B" -> {b + 1, w}
          "W" -> {b, w + 1}
          _ -> {b, w}
        end
      end)
    end)
    |> case do
      {0, 0} -> :none
      {_, 0} -> :black
      {0, _} -> :white
      _ -> :none
    end
  end

  defp to_groups(g) do
    g
    |> Enum.reduce({[], %MapSet{}}, fn
      {_, c}, acc when c != "_" ->
        acc

      {pos, _}, {groups, visited} ->
        {group, visited} =
          pos
          |> :queue.in(:queue.new())
          |> bfs(visited, g, [])

        {[group | groups], visited}
    end)
    |> elem(0)
  end

  defp to_grid(board) do
    board
    |> String.split("\n")
    |> Enum.with_index()
    |> Enum.reduce(%{}, fn {r, i}, acc ->
      r
      |> String.graphemes()
      |> Enum.with_index()
      |> Enum.reduce(acc, fn {c, j}, acc ->
        Map.put(acc, {j, i}, c)
      end)
    end)
  end

  defp bfs(q, visited, g, group) do
    case :queue.out(q) do
      {:empty, _} ->
        {Enum.sort(group), visited}

      {{:value, pos = {x, y}}, q} ->
        cond do
          pos in visited ->
            bfs(q, visited, g, group)

          g[pos] != "_" ->
            bfs(q, visited, g, group)

          true ->
            visited = MapSet.put(visited, pos)
            group = [pos | group]

            [{-1, 0}, {0, 1}, {1, 0}, {0, -1}]
            |> Enum.reduce(q, fn {dx, dy}, q ->
              :queue.in({x + dx, y + dy}, q)
            end)
            |> bfs(visited, g, group)
        end
    end
  end

  @doc """
  Return all white, black and neutral territories
  """
  @spec territories(board :: String.t()) :: territories
  def territories(board) do
    g = board |> to_grid()

    g
    |> to_groups()
    |> Enum.reduce(%{black: [], none: [], white: []}, fn t, acc ->
      k = check(t, g)
      Map.update!(acc, k, &(&1 ++ t))
    end)
    |> Map.new(fn {k, v} -> {k, Enum.sort(v)} end)
  end
end
